### PR TITLE
[mbedTLS] Enable TLS 1.3 support

### LINKED ISSUE #92101: WebSocketPeer uses TLSv1.2 and can't change to TLSv1.3
### Tested versions

v4.2.1.stable.official [15073afe3] 
...
Godot_v4.3-dev6_linux

### System information

Godot v4.2.2.stable - Ubuntu 23.10 23.10 - X11 - GLES3 (Compatibility) - llvmpipe (LLVM 15.0.7, 256 bits) () - AMD Ryzen 9 5900X 12-Core Processor (24 Threads)

### Issue description

image

I'm connecting to a WSS source from Godot. Wireshark shows that it uses TLSv1.2 protocol.
The connection is STATE_OPEN, but packets are not sent from server to me.

When I'm testing the same WSS address in Postman everything works perfectly and Wireshark shows that TLSv1.3 is used. Also there are some other differences.

Also, the documentation is absolutely useless. So nobody knows if this is a Godot bug, or just outdated version or something that is possible to fix in code, but unclear how!

I asked about the problem 4 times in discord and literally 0 answers.

### Steps to reproduce

connect to any websocket that is uses TLSv1.3

### Minimal reproduction project (MRP)

[Archive.zip](https://github.com/godotengine/godot/files/15367289/Archive.zip)

For some unclear reason this MRP can't even connect with 100% the same code as original. 
The original can at least connect, although doesn't receive anything

### PR DESCRIPTION
**Pull request by Faless**
Move library initialization to module registration functions.

Only set library debug threshold when verbose output is enabled.

TLSv1.3 functions seems to be a bit more verbose then expected, and generate a lot of noise. Yet, some level of debugging without recompiling the engine would be nice. We should discuss this upstream.

Draft status:
- ~Depends on #96385~
- ~Depends on #97458~

Fixes #92101

Test:

```gdscript
# main.gd
extends Node

var http : HTTPRequest

func _ready() -> void:
	http = HTTPRequest.new()
	add_child(http)
	http.request_completed.connect(_completed)
	if http.request("https://tls13.akamai.io/") != OK:
		_exit()

func _completed(result: int, code: int, headers: PackedStringArray, body: PackedByteArray):
	if code != 200:
		print("Request failed, result: %d, code: %d" % [result, code])
		_exit()
		return
	var str_body := body.get_string_from_utf8()
	var tls13 := "Your client negotiated TLS 1.3" in str_body
	var tls12 := "Your client negotiated TLS 1.2" in str_body
	if tls13:
		print("TLS 1.3 supported :)")
	elif tls12:
		print("TLS 1.3 **NOT** supported :(")
		print("TLS 1.2 supported :)")
	else:
		print("Can't determine TLS version... Check yourself: %s" % str_body)
	_exit()

func _exit():
	get_tree().quit.call_deferred()
```

Note for package maintainers: When using `builtin_mbedtls=no` TLS 1.3 support will depend on the system library version and configuration.

### CODE REVIEW

**Comment by fire** [comment]
The cicd check is failing on `The command line is too long.`.

**Comment by Faless** [comment]
> The cicd check is failing on `The command line is too long.`.

Yes, seems like a build system issue on Windows + MSVC (not MinGW apparently).

I don't know how to fix that and I currently can't easily test MSVC builds, so help is welcome.

**Comment by akien-mga** [comment]
> > The cicd check is failing on `The command line is too long.`.
> 
> Yes, seems like a build system issue on Windows + MSVC (not MinGW apparently).
> 
> I don't know how to fix that and I currently can't easily test MSVC builds, so help is welcome.

CC @bruvzg @Repiteo

**Comment by bruvzg** [comment]
> I don't know how to fix that and I currently can't easily test MSVC builds, so help is welcome.

Probably in the same way as `ar` - https://github.com/godotengine/godot/pull/96407, MSVC `lib` do support running with response file - https://learn.microsoft.com/en-us/cpp/build/reference/running-lib?view=msvc-170#lib-command-files

**Comment by Faless** [comment]
> Probably in the same way as `ar` - #96407, MSVC `lib` do support running with response file

I remember trying it but it didn't work. I added https://github.com/godotengine/godot/pull/96394/commits/3c66820cf375b45b3dd14aa6ade1575b188e385f and indeed it's still failing (doesn't seem to use the tempfile).

Am I doing something wrong?

**Comment by bruvzg** [comment]
> Am I doing something wrong?

It's actually already is using `TEMPFILE` by default, the issue is `env["MAXLINELENGTH"] = 8192` line in our config (added in #87154), removing it fixes the build.

**Comment by bruvzg** [comment]
For the reference, max length seems to be `8191` - https://learn.microsoft.com/en-us/troubleshoot/windows-client/shell-experience/command-line-string-limitation#more-information, but it's not working either. `8000` seems to working, so not sure what's the real limit is. Scons default is `2048`.

**Comment by Faless** [comment]
Opened #97458 with the windows fix.

**Comment by DjSapsan** [comment]
OMG you are hero!
I'm testing this change right now and it's already looks promising.
As it's just a commit I needed to compile Godot from scratch, so not sure if everything will work 100% but I hope all good.

**Review by fire — APPROVED** [review]
I expect small tests to work but I don’t think anyone in the godot developer community would be able to test the obscure browser combinations.

My vote is to send the branch into to a pre-release asap.

Edited:

Since @Faless is the web maintainer I don’t know who else can really review it.

**Comment by DjSapsan** [comment]
I tested it in my project. My application was finally loading data for a period but then started failing.
Sadly, I can't 100% confirm why it worked.
After it started failing I checked Wireshark and it's showing TLS 1.2 again. Not sure if it used 1.3 during successes.
No amount of reboots and code changes make it 1.3.

File: modules/mbedtls/register_types.cpp [hunk 1]
```diff
@@ -35,15 +35,34 @@
...
 
+#if MBEDTLS_VERSION_MAJOR >= 3
+#include <psa/crypto.h>
+#endif
+
 #ifdef TESTS_ENABLED
 #include "tests/test_crypto_mbedtls.h"
 #endif
 
+static bool godot_mbedls_initialized = false;
```
Comment (akien-mga) [inline modules/mbedtls/register_types.cpp:46]:
```suggestion
static bool godot_mbedtls_initialized = false;
```

**Review by akien-mga — APPROVED** [review]
Tested on Linux, seems to work for me:
```
TLS 1.3 supported :)
```

**Review by akien-mga — APPROVED** [review]
Good catch to update the docs too.

**Comment by akien-mga** [comment]
Thanks!

**Comment by DjSapsan** [comment]
Testing again (Kubuntu btw).
It's showing the correct TLS 1.3, though I still have problems with loading data.
After first packet it stops refreshing

**Comment by DjSapsan** [comment]
I tested the issue deeper with Wireshark. I need to figure out if this problem with my code or the Godot's code.
I removed everything from my code and just receiving packets without any computation. 
After a short working period it shows errors of ZeroWindow and sometimes others. It means that the client (Godot app) can't receive data fast enough. The Browser and Postman still loading the same data perfectly fine.
I think it has something to do with the buffers or maybe something else.
image

**Comment by Faless** [comment]
@DjSapsan please open a dedicated issue with a minimal reproduction project. It's not really possible to understand what's going on otherwise.

**Comment by DjSapsan** [comment]
> @DjSapsan please open a dedicated issue with a minimal reproduction project. It's not really possible to understand what's going on otherwise.

https://github.com/godotengine/godot/issues/97899

**Pull request merged**